numpy/ma/tests/test_core.py

```diff
@@ -5017,6 +5017,28 @@ def test_astype():
     assert_(x_f2.mask.flags.f_contiguous)
 
 
+# This sort of test over types could be done with a fixture
+def test_astype_basic():
+    # See gh-12070
+    dt = [np.dtype(dt_) for dt_ in '?bhilqpBHILQPefdgFD']
+    with suppress_warnings() as sup:
+        sup.filter(np.ComplexWarning)
+        for dt1, dt2 in itertools.product(dt, dt):
+            src = np.ma.array(ones(3, dt1), fill_value=1)
+            # Finalize the fill_value for structured arrays
+            dst = src.astype(dt2)
+
+            assert_(src.fill_value == 1)
+            assert_(src.dtype == dt1)
+            assert_(np.dtype(src.fill_value) == dt1)
+
+            assert_(dst.fill_value == 1)
+            assert_(dst.dtype == dt2)
+            assert_(np.dtype(dst.fill_value) == dt2)
```

Comment:
np.dtype doesn't work liked this, does it?

Reply:
Seems to work here, I'll change it as the method looks a bit nicer.